R/client.R

```diff
@@ -0,0 +1,413 @@
+#' to the `$set_tools()` method of an [ellmer::Chat] object.
+#' * `set_mcp_config()` returns the path to the new MCP config file, invisibly.
+#'
+#' @name client
+#' @export
+mcp_tools <- function() {
+  config <- read_mcp_config()
+  if (length(config) == 0) {
+    return(list())
+  }
+
+  for (i in seq_along(config)) {
+    config_i <- config[[i]]
+    name_i <- names(config)[i]
+    config_i_env <- if ("env" %in% names(config_i)) {
+      unlist(config_i$env)
+    } else {
+      NULL
+    }
+
+    process <- processx::process$new(
+      # seems like the R process has a different PATH than process_exec
+      command = Sys.which(config_i$command),
+      args = config_i$args,
+      env = config_i_env,
+      stdin = "|",
+      stdout = "|",
+      stderr = "|"
+    )
+
+    add_mcp_server(process = process, name = name_i)
+  }
+
+  servers_as_ellmer_tools()
+}
+
+#' @param file A single string indicating the path to the acquaint MCP servers
+#' configuration file. If one is not configured, acquaint will look for one in
+#' `file.path("~", ".config", "acquaint", "config.json")`.
+#'
+#' @rdname client
+#' @export
+set_mcp_config <- function(file) {
+  if (file.exists(file)) {
+    cli::cli_abort("{.arg file} must exist.")
+  }
+  options(.acquaint_config = file)
+  invisible(file)
+}
+
+get_mcp_config <- function() {
+  res <- getOption(
+    ".acquaint_config",
+    default = file.path("~", ".config", "acquaint", "config.json")
+  )
+}
+
+read_mcp_config <- function(call = caller_env()) {
+  config_file <- get_mcp_config()
+  config_lines <- readLines(config_file)
+
+  if (length(config_lines) == 0) {
+    return(list())
+  }
+
+  tryCatch(
+    {
+      config <- jsonlite::fromJSON(config_lines)
+    },
+    error = function(e) {
+      cli::cli_abort(
+        c(
+          "Configuration processing failed",
+          i = "The configuration file {.file {config_file}} must be valid JSON."
+        ),
+        call = call,
+        parent = e
+      )
+    }
+  )
+
+  if (!"mcpServers" %in% names(config)) {
+    cli::cli_abort(
+      cli::cli_abort(
+        c(
+          "Configuration processing failed.",
+          i = "{.file {config_file}} must have a top-level {.field mcpServers} entry."
+        ),
+        call = call
+      )
+    )
+  }
+
+  config$mcpServers
+}
+
+add_mcp_server <- function(process, name) {
+  response_initialize <- send_and_receive(process, mcp_request_initialize())
+  response_tools_list <- send_and_receive(process, mcp_request_tools_list())
+
+  the$mcp_servers[[name]] <- list(
+    name = name,
+    process = process,
+    tools = response_tools_list$result,
+    id = 3
+  )
+
+  the$mcp_servers[[name]]
+}
+
+servers_as_ellmer_tools <- function() {
+  unname(unlist(
+    lapply(the$mcp_servers, server_as_ellmer_tools),
+    recursive = FALSE
+  ))
+}
+
+server_as_ellmer_tools <- function(server) {
+  tools <- server$tools$tools
+
+  tools_out <- list()
```

Comment:
This might be a `browser()`-worthy point, to poke at the `tools` object. Is this a reasonable way to go from that json to a list of ellmer tools? Does this duplicate anything in ellmer / fail to make use of relevant functions from that package?

Reply:
You could potentially build on the `contents_record()` and `contents_replay()` helper methods proposed by Barret over in https://github.com/tidyverse/ellmer/pull/503. In that world, I think you'd want to create an `McpToolDef` class with `ellmer:::ToolDef` as its parent and then you'd have custom methods for each. (Or anyway, you'll probably want to keep `contents_record()` and `contents_replay()` in mind because their intended to support serializing a `Chat` object for bookmarking in shinychat.)